hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/BlockOutputStreamEntry.java

```diff
@@ -113,8 +113,13 @@ public void write(int b) throws IOException {
   @Override
   public void write(byte[] b, int off, int len) throws IOException {
     checkStream();
-    getOutputStream().write(b, off, len);
-    incCurrentPosition(len);
+    try {
+      getOutputStream().write(b, off, len);
+    }finally {
+      //Even though actual write failed we will move the position and when
```

Comment:
I am really not sure about this statement...
The first things comes to my mind is: If so, then we should do similarly in the write(int) method, and possibly in writeOnRetry as well.

A second point I am not sure about is that, resetToAckedDataLength is the method that sets back the position to the acked position. In KeyOutputStream it is called from handleException, but there are not other calls to the reset method and ECKeyOutputStream's handleException does not call the reset. What is the mechanism I don't see now, that resets the position in case of EC? As ECBlockOutputStreamEntry relies on these same write methods.

Reply:
From *KeyOututStream: write(int) will use  write(buf, 0, 1); method only.
```
  @Override
  public void write(int b) throws IOException {
    byte[] buf = new byte[1];
    buf[0] = (byte) b;
    write(buf, 0, 1);
  }
```
I am not very sure how we are using write(int) in *BlockOutputStream itself. This is not exposed impl.

We are resetting the length in handleStripeFailure. HandleException is just setting the IOException when writing. After and before putBlock we will verify the failure and reset the lengths on the streams as needed if there are failures.

Reply:
Thank you for the explanation, now I think I see why it is enough to do this change just here and still remain correct.
So:
- write(int) is not used, no problem with how it is implemented
- writeOnRetry is ok to not increase the current position on an error, as it is not called in EC case so far, and during handleException before writeOnRetry we reset the position with resetToAckedDataLength in non-EC case.
This basically means that if we would increase first, then write to stream, the error handling code will possibly always reset (it depends on how we implement multiple retries in EC case).

I think it took me quite a time to understand and for you to explain why this is correct.

I see two ways out from this which makes any of our predecessors to understand the code a bit easier:
1. increase the current position before writing to the underlying stream in all three methods for clarity, and let error handling reset if there is an error
2. implement a logic to have acked data length persisted in the EC case, and use the reset method there as well, instead of decrementing current position by the last stripe size, so that we do not need to ensure that the increase happens.

Based on my understanding, I think leaving it just this way would work, but would not help to get what the code does later on when any of us has to come back and debug or change something if we leave the code this way, and we have methods that increment the current position differently.

Reply:
Now we don't need this changes in BlockOutputStreamEntry as we are tracking last successful stripe position.